File: packages/mesofact-build/src/bundle.ts
//! @yah:ticket(R012-T1, "Mode 3 build: client-tree bundling → dist/hydrate/, shell prerender with __MESOFACT_STATE__ + hydrate-script injection, manifest hydration population")
//! @yah:at(2026-05-26T16:04:16Z)
//! @yah:status(review)
//! @yah:assignee(agent:claude)
//! @yah:phase(P10)
//! @yah:parent(R012)
//! @yah:handoff("Mode 3 build tree shipped end-to-end (TS). (1) bundle.ts: bundleClientEntrypoints() bundles each spa route's client_entrypoint to dist/hydrate/ — browser target, ESM, splitting:true, content-hashed names (entry '<key>.[hash].js', chunks '<key>.chunk-[hash].js'). Returns {script, code_split} (basenames). (2) routes.ts: RouteEntry gains client_entrypoint?; build throws if a spa route omits it or a non-spa route sets it. (3) manifest-build.ts: AssembleInput.hydration map (route→ManifestHydration) populates ManifestRoute.hydration for spa routes. (4) prerender.ts: now prerenders static AND spa (skips ssr); for spa inputs it weaves the shell — serializeState() HTML-escapes initial_state (<,>,&,U+2028/9) into <script id=\"__MESOFACT_STATE__\" type=\"application/json\">, then appends <script type=module src=/{build_id}/hydrate/{script}> before </body>. (5) contract.ts: Hydration.script relaxed to optional (build owns the hashed script/code_split; render ships initial_state) + documents the 6-line client snippet. index.ts wires client bundling + hydration map + spa prerender inputs.")
//! @yah:verify("cd packages/mesofact-runtime && bun run typecheck && bun run build && bun test")
//! @yah:verify("cd packages/mesofact-build && bun run typecheck && bun test")
//! @yah:assumes("Code-split chunks are fetched by the entry script's own relative ESM imports once it's served from /{build_id}/hydrate/, so the shell only injects the entry <script>; the manifest still lists code_split so the publisher uploads every chunk.")
//! @yah:assumes("Runtime dist must be rebuilt (bun run build in mesofact-runtime) for build/worker to see the new client_entrypoint + optional Hydration.script types — @mesofact/runtime resolves via dist/, not src/.")
//!
//! @yah:ticket(R015-F4, "Server/client module boundary lint — host-only API import detection")
//! @yah:assignee(agent:claude)
//! @yah:at(2026-06-04T19:32:49Z)
//! @yah:status(review)
//! @yah:phase(P2)
//! @yah:parent(R015)
//! @yah:next("Build-time walk of the module graph reachable from client_entrypoint. Imports of host-only APIs (configurable list: node:fs, node:net, named db drivers, etc.) fail the build with the offending import chain.")
//! @yah:next("Same machinery also enforces W173 Placement validation: reject placement:\"host\" on a route importing edge-only APIs, and placement:\"edge\" on a route importing host-only APIs.")
//! @yah:next("Implementation: spike Bun.build's artifact `imports` analysis vs `bun build --metafile=path` (esbuild-shape JSON) and pick before shipping. The bundler is already Bun.build (see bundle.ts) so artifact analysis is the lower-friction path if it exposes the import graph.")
//! @yah:next("Soft-gated on R434-F5 (yah-side first SSR consumer) for end-to-end validation, but the SPA-side lint (host-only APIs in a spa client_entrypoint) is testable today without it.")
//! @yah:next("Throwaway when RSC's \"use client\" lands; same idea, finer granularity.")
//! @yah:verify("A test fixture client_entrypoint that imports node:fs fails the build with the import chain in the error")
//! @yah:verify("A test fixture ssr+placement:\"edge\" entrypoint that imports a db driver fails the build")
//! @yah:verify("Existing yah-side SPA routes (../../app/yah/web/marketing, ../../app/yah/web/dashboard) build clean")
//! @yah:handoff("Server/client module boundary lint shipped. New packages/mesofact-build/src/host-lint.ts exports BROWSER_FORBIDDEN (node:* regex + bare-name builtins fs/path/net/...) and EDGE_FORBIDDEN (browser list + db drivers pg/mysql2/mongodb/redis/ioredis/better-sqlite3). Implementation runs Bun.build with a side-effect onResolve plugin and no outdir — zero files hit disk, full transitive import graph captured. If a forbidden specifier doesn't resolve at all (e.g. `pg` not installed), Bun.build throws Bundle failed; we swallow that and surface the recorded violation instead, since the lint error names the offending importer chain (much more actionable than 'Bundle failed'). Wired into build pipeline BEFORE bundleEntrypoints so violations short-circuit the cascade rather than tripping a downstream bundler error. Two fixtures: spa-host-only (client_entrypoint → describe.ts → node:fs — confirms transitive walk + chain in error message) and ssr-edge-host-only (ssr placement:'edge' importing pg). Existing ssr fixture confirms placement:'host' is permissive (no false positive). Verified on real yah-side consumers: app/yah/web/marketing (5 routes, 3 spa) and app/yah/web/dashboard (7 routes, 6 spa) both `bun run build` clean. Test totals: mesofact-build 33 pass (was 30); mesofact-runtime 57 pass; typecheck clean across runtime/build/worker. Configurable-list extensibility (W173 'configurable list') deferred until a real override case surfaces — the named exports BROWSER_FORBIDDEN/EDGE_FORBIDDEN can be re-exported or overridden via a BuildOptions field later without API churn.")
//! @yah:verify("cd packages/mesofact-build && bun test — 33 pass")
//! @yah:verify("cd packages/mesofact-build && bun run typecheck — clean")
//! @yah:verify("cd app/yah/web/marketing && bun run build — clean")
//! @yah:verify("cd app/yah/web/dashboard && bun run build — clean")

// Phase 1 — bundle each route's server entrypoint to ESM under `dist/server/`.
// Uses Bun's bundler (Bun.build) which natively understands TS, JSX, and
// workspace imports. The Vite/Bun split mentioned in the architecture is a
// future concern (Mode 3 client tree); P5 only needs the server tree.
//
// Each entrypoint becomes a deterministic output filename
// (`<route_key>.js`) so the manifest's `render_entrypoint` can name a stable
// path regardless of source layout. `route_key` comes from the route pattern
// (see `route-key.ts`).
//
// External dep: `@mesofact/runtime` is kept external — the build process
// already has it loaded, and bundling it would duplicate the
// `AsyncLocalStorage` registry that backs `runInTrackCtx`. The publisher
// (P6) zips the entrypoint + node_modules as a deployable unit.

import { mkdir } from "node:fs/promises";
import { basename, join, resolve } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { BuildError } from "./load-routes.js";
import { routeKey } from "./route-key.js";

// SSR routes export a Web Fetch handler as `default` —
// `(req: Request) => Promise<Response>`. Static/spa routes export a `render`
// fn (validated lazily by the prerender pass). The Fetch signature can't be
// proven from the module's static shape — only that `default` is callable —
// so this verifies the call shape and leaves runtime behavior to the dev
// proxy / pond container / Worker. Catches the common authoring slip:
// shipping `render` for a `mode:"ssr"` route, or no default export at all.
export async function assertSsrEntrypoint(route: string, bundlePath: string): Promise<void> {
  let mod: { default?: unknown };
  try {
    mod = (await import(pathToFileURL(bundlePath).href)) as { default?: unknown };
  } catch (e) {
    throw new BuildError(
      `route ${route}: failed to import SSR bundle at ${bundlePath}: ${(e as Error).message}`,
    );
  }
  if (typeof mod.default !== "function") {
    throw new BuildError(
      `route ${route}: mode:"ssr" entrypoint must \`export default\` a Fetch handler ` +
        `\`(req: Request) => Promise<Response>\` (got ${describe(mod.default)})`,
    );
  }
}

function describe(v: unknown): string {
  if (v === undefined) return "no default export";
  if (v === null) return "null";
  return typeof v;
}

export type BundleInput = {
  route: string;
  entrypoint: string; // path resolved against `projectRoot`
};

export type BundleOutput = {
  route: string;
  // Path written under `outDir/server/`. Used as the manifest's
  // `render_entrypoint`.
  serverPath: string;
  // Absolute path for the prerender driver to dynamic-import.
  absolutePath: string;
};

export async function bundleEntrypoints(
  projectRoot: string,
  outDir: string,
  inputs: readonly BundleInput[],
): Promise<BundleOutput[]> {
  if (inputs.length === 0) return [];
  const serverDir = join(outDir, "server");
  await mkdir(serverDir, { recursive: true });

  // Bun.build emits one file per entrypoint, named after the source basename.
  // We want a deterministic name per route — bundle one at a time and rename
  // via `naming`. (Bun's naming token `[name]` defaults to source basename;
  // setting it explicitly per-call gives us the route_key shape we want.)
  const outputs: BundleOutput[] = [];
  for (const input of inputs) {
    const key = routeKey(input.route);
    const absEntry = resolve(projectRoot, input.entrypoint);
    const result = await Bun.build({
      entrypoints: [absEntry],
      outdir: serverDir,
      target: "bun",
      format: "esm",
      naming: `${key}.js`,
      external: ["@mesofact/runtime"],
      splitting: false,
      sourcemap: "none",
    });
    if (!result.success) {
      const msg = result.logs.map((l) => l.message).join("\n");
      throw new BuildError(`bundle failed for route ${input.route}: ${msg}`);
    }
    const written = result.outputs.find((o) => o.kind === "entry-point");
    if (!written) {
      throw new BuildError(`bundle for route ${input.route} produced no entry-point output`);
    }
    const absolutePath = fileURLToPath(pathToFileURL(written.path).href);
    outputs.push({
      route: input.route,
      serverPath: `dist/server/${key}.js`,
      absolutePath,
    });
  }
  return outputs;
}

// ─── Mode 3 client tree ───────────────────────────────────────────────────────

export type ClientBundleInput = {
  route: string;
  clientEntrypoint: string; // path resolved against `projectRoot`
};

export type ClientBundleOutput = {
  route: string;
  // Entry-point filename, content-hashed, relative to `dist/hydrate/`. Goes in
  // the manifest's `hydration.script` and is referenced by the shell's
  // `<script type="module" src="/{build_id}/hydrate/<script>">`.
  script: string;
  // Code-split chunk filenames (also under `dist/hydrate/`). The entry imports
  // them by relative path, so the browser fetches them from the same prefix;
  // the manifest lists them so the publisher uploads every chunk.
  code_split: string[];
};

// Phase 1b — bundle each Mode 3 route's *client* entrypoint to `dist/hydrate/`.
// Unlike the server tree, the client tree targets the browser, content-hashes
// outputs (immutable CDN caching, see §"Static asset handling"), and enables
// code splitting so shared chunks are emitted once. `@mesofact/runtime` is NOT
// external here — the client never touches the server-only adapter registry;
// any runtime imports in a client entry must be type-only (erased at build).
export async function bundleClientEntrypoints(
  projectRoot: string,
  outDir: string,
  inputs: readonly ClientBundleInput[],
): Promise<ClientBundleOutput[]> {
  if (inputs.length === 0) return [];
  const hydrateDir = join(outDir, "hydrate");
  await mkdir(hydrateDir, { recursive: true });

  const outputs: ClientBundleOutput[] = [];
  for (const input of inputs) {
    const key = routeKey(input.route);
    const absEntry = resolve(projectRoot, input.clientEntrypoint);
    const result = await Bun.build({
      entrypoints: [absEntry],
      outdir: hydrateDir,
      target: "browser",
      format: "esm",
      splitting: true,
      sourcemap: "none",
      naming: {
        entry: `${key}.[hash].js`,
        chunk: `${key}.chunk-[hash].js`,
        asset: "[name].[hash].[ext]",
      },
    });
    if (!result.success) {
      const msg = result.logs.map((l) => l.message).join("\n");
      throw new BuildError(`client bundle failed for route ${input.route}: ${msg}`);
    }
    const entry = result.outputs.find((o) => o.kind === "entry-point");
    if (!entry) {
      throw new BuildError(
        `client bundle for route ${input.route} produced no entry-point output`,
      );
    }
    const code_split = result.outputs
      .filter((o) => o.kind === "chunk")
      .map((o) => basename(o.path))
      .sort();
    outputs.push({
      route: input.route,
      script: basename(entry.path),
      code_split,
    });
  }
  return outputs;
}

File: crates/mesofact-build/src/bundle.rs
//! Rolldown bundling (W174 pillar 2 / R448). Mirrors
//! `packages/mesofact-build/src/bundle.ts`:
//!
//! - server tree: one bundle per route entrypoint → `dist/server/<key>.js`,
//!   ESM, no splitting, `@mesofact/runtime` external.
//! - client tree: one bundle per client entrypoint → `dist/hydrate/
//!   <key>.<hash>.js` (+ `<key>.chunk-<hash>.js` shared chunks), browser
//!   platform, content-hashed.
//!
//! Divergence from the Bun pipeline, by design: server bundles resolve with
//! browser-ish conditions (Platform::Browser) instead of Bun's node-flavored
//! target, because the SSG/SSR executor is deno_core (workerd-semantics V8)
//! — react-dom must resolve to `server.browser` rather than the
//! node-streams build. The Fetch-handler contract is identical either way
//! (W173 § "Entrypoint signatures"); the W174 amendment records this.

use anyhow::{anyhow, bail, Result};
use rolldown::{
    Bundler, BundlerOptions, ChunkFilenamesOutputOption, InputItem, IsExternal, OutputFormat,
    Platform,
};
use rolldown_common::Output;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use crate::route_key::route_key;

pub struct ServerBundle {
    pub route: String,
    /// Manifest `render_entrypoint` value (`dist/server/<key>.js`).
    pub server_path: String,
    /// Absolute path for the SSG executor to import.
    pub absolute_path: PathBuf,
    /// Module ids that ended up in the bundle (for the host-lint pass).
    pub module_ids: Vec<String>,
    /// Chunk-level `imports` (cross-chunk filenames + external module
    /// specifiers, e.g. `node:fs`) for the host-lint pass. Rolldown
    /// externalizes `node:*`/bare builtins rather than bundling them, and
    /// externalized ids are NOT included in `module_ids` — this is where
    /// they show up instead (R513-B11). See `assert_no_forbidden_modules`.
    pub import_ids: Vec<String>,
}

pub struct ClientBundle {
    pub route: String,
    /// Content-hashed entry filename relative to `dist/hydrate/`.
    pub script: String,
    /// Code-split chunk filenames (sorted), also under `dist/hydrate/`.
    pub code_split: Vec<String>,
    pub module_ids: Vec<String>,
    /// See `ServerBundle::import_ids` — same rationale, aggregated across
    /// every chunk this client bundle emitted (entry + code-split).
    pub import_ids: Vec<String>,
}

fn base_options(project_root: &Path) -> BundlerOptions {
    BundlerOptions {
        cwd: Some(project_root.to_path_buf()),
        format: Some(OutputFormat::Esm),
        // Renders execute with production React; the Bun pipeline relies on
        // Bun's own NODE_ENV inlining for browser targets — pin it here so
        // both pipelines drop the dev branches.
        define: Some(
            [(
                "process.env.NODE_ENV".to_string(),
                "\"production\"".to_string(),
            )]
            .into_iter()
            .collect(),
        ),
        ..Default::default()
    }
}

async fn run_bundler(options: BundlerOptions, what: &str) -> Result<Vec<Output>> {
    let mut bundler =
        Bundler::new(options).map_err(|e| anyhow!("{what}: bundler init failed: {e:?}"))?;
    let result = bundler.write().await;
    let output = match result {
        Ok(o) => o,
        Err(errs) => bail!("{what}: bundle failed: {errs:?}"),
    };
    bundler.close().await.map_err(|e| anyhow!("{what}: bundler close failed: {e:?}"))?;
    Ok(output.assets)
}

/// Bundle each route's server entrypoint to `dist/server/<key>.js`. One
/// bundler invocation per route — deterministic names, no cross-route chunk
/// sharing (parity with the Bun pipeline's per-route `Bun.build` calls).
pub async fn bundle_server_entrypoints(
    project_root: &Path,
    out_dir: &Path,
    inputs: &[(String, String)], // (route, entrypoint)
) -> Result<Vec<ServerBundle>> {
    let server_dir = out_dir.join("server");
    std::fs::create_dir_all(&server_dir)?;

    let mut outputs = Vec::new();
    for (route, entrypoint) in inputs {
        let key = route_key(route);
        let abs_entry = project_root.join(entrypoint);
        let options = BundlerOptions {
            input: Some(vec![InputItem {
                name: Some(key.clone()),
                import: abs_entry.to_string_lossy().into_owned(),
            }]),
            dir: Some(server_dir.to_string_lossy().into_owned()),
            platform: Some(Platform::Browser),
            external: Some(IsExternal::from(vec!["@mesofact/runtime".to_string()])),
            entry_filenames: Some(ChunkFilenamesOutputOption::String("[name].js".to_string())),
            ..base_options(project_root)
        };
        let assets = run_bundler(options, &format!("route {route}")).await?;
        let entry = assets
            .iter()
            .find_map(|a| match a {
                Output::Chunk(c) if c.is_entry => Some(c),
                _ => None,
            })
            .ok_or_else(|| anyhow!("bundle for route {route} produced no entry-point output"))?;
        outputs.push(ServerBundle {
            route: route.clone(),
            server_path: format!("dist/server/{key}.js"),
            absolute_path: server_dir.join(entry.filename.as_str()),
            module_ids: entry.module_ids.iter().map(|m| m.to_string()).collect(),
            import_ids: entry.imports.iter().map(|m| m.to_string()).collect(),
        });
    }
    Ok(outputs)
}

/// Bundle each client entrypoint to `dist/hydrate/<key>.<hash>.js` with
/// code splitting. `@mesofact/runtime` is NOT external here (parity with
/// bundle.ts — client code must not touch the server adapter registry).
pub async fn bundle_client_entrypoints(
    project_root: &Path,
    out_dir: &Path,
    inputs: &[(String, String)], // (route, client_entrypoint)
) -> Result<Vec<ClientBundle>> {
    let hydrate_dir = out_dir.join("hydrate");
    if inputs.is_empty() {
        return Ok(Vec::new());
    }
    std::fs::create_dir_all(&hydrate_dir)?;

    let mut outputs = Vec::new();
    for (route, client_entrypoint) in inputs {
        let key = route_key(route);
        let abs_entry = project_root.join(client_entrypoint);
        let options = BundlerOptions {
            input: Some(vec![InputItem {
                name: Some(key.clone()),
                import: abs_entry.to_string_lossy().into_owned(),
            }]),
            dir: Some(hydrate_dir.to_string_lossy().into_owned()),
            platform: Some(Platform::Browser),
            entry_filenames: Some(ChunkFilenamesOutputOption::String(format!("{key}.[hash].js"))),
            chunk_filenames: Some(ChunkFilenamesOutputOption::String(format!("{key}.chunk-[hash].js"))),
            ..base_options(project_root)
        };
        let assets = run_bundler(options, &format!("client bundle for route {route}")).await?;
        let mut script = None;
        let mut code_split = Vec::new();
        let mut module_ids = Vec::new();
        let mut import_ids = Vec::new();
        for a in &assets {
            if let Output::Chunk(c) = a {
                module_ids.extend(c.module_ids.iter().map(|m| m.to_string()));
                import_ids.extend(c.imports.iter().map(|m| m.to_string()));
                if c.is_entry {
                    script = Some(c.filename.to_string());
                } else {
                    code_split.push(c.filename.to_string());
                }
            }
        }
        code_split.sort();
        let script = script
            .ok_or_else(|| anyhow!("client bundle for route {route} produced no entry-point output"))?;
        outputs.push(ClientBundle { route: route.clone(), script, code_split, module_ids, import_ids });
    }
    Ok(outputs)
}

/// Bundle the routes file itself (`mesofact.routes.ts`) for evaluation in
/// the SSG runtime. `@mesofact/runtime` stays external (the executor maps it
/// to the embedded shim); everything else the file imports is bundled in.
pub async fn bundle_routes_file(project_root: &Path, scratch_dir: &Path) -> Result<PathBuf> {
    let routes_file = project_root.join("mesofact.routes.ts");
    if !routes_file.exists() {
        bail!("expected {} to exist", routes_file.display());
    }
    std::fs::create_dir_all(scratch_dir)?;
    let options = BundlerOptions {
        input: Some(vec![InputItem {
            name: Some("routes".to_string()),
            import: routes_file.to_string_lossy().into_owned(),
        }]),
        dir: Some(scratch_dir.to_string_lossy().into_owned()),
        platform: Some(Platform::Neutral),
        external: Some(IsExternal::from(vec!["@mesofact/runtime".to_string()])),
        entry_filenames: Some(ChunkFilenamesOutputOption::String("[name].mjs".to_string())),
        ..base_options(project_root)
    };
    run_bundler(options, "mesofact.routes.ts").await?;
    Ok(scratch_dir.join("routes.mjs"))
}

/// Map of forbidden import → reason, mirroring host-lint.ts's
/// BROWSER_FORBIDDEN / EDGE_FORBIDDEN sets.
pub fn browser_forbidden(module_id: &str) -> bool {
    const BARE_BUILTINS: &[&str] = &[
        "fs", "path", "net", "os", "child_process", "worker_threads", "cluster", "dgram",
        "dns", "tls", "http", "https", "http2", "stream", "zlib", "crypto", "vm",
    ];
    module_id.starts_with("node:")
        || BARE_BUILTINS.contains(&module_id)
        || module_id.starts_with("bun:")
}

pub fn edge_forbidden(module_id: &str) -> bool {
    const DB_DRIVERS: &[&str] = &["pg", "mysql2", "mongodb", "redis", "ioredis", "better-sqlite3"];
    browser_forbidden(module_id)
        || DB_DRIVERS.iter().any(|d| module_id == *d || module_id.starts_with(&format!("{d}/")))
}

/// Post-bundle lint over the captured module graph. The TS pipeline walks
/// the graph with an onResolve plugin pre-bundle; here the bundle already
/// ran (rolldown reports unresolved forbidden ids as bundle errors, which
/// surface first with the same actionable specifier), and resolved-but-
/// forbidden ids are caught here.
///
/// Two sources are scanned, not one (R513-B11): `module_ids` covers
/// resolved-and-bundled files, but rolldown externalizes `node:*` and other
/// bare builtins instead of bundling them — an externalized
/// `import "node:fs"` never appears in `module_ids`, only in the chunk's
/// `imports` list (`import_ids` here). A hydrate bundle that ships a
/// side-effectful `node:*` import used to sail through this lint untouched.
pub fn assert_no_forbidden_modules(
    route: &str,
    kind: &str,
    module_ids: &[String],
    import_ids: &[String],
    forbidden: fn(&str) -> bool,
) -> Result<()> {
    // module ids are absolute paths for resolved files; bare/builtin ids
    // stay as written (external or shimmed). import_ids mixes cross-chunk
    // filenames (never forbidden) with external specifiers (may be).
    let offenders: BTreeMap<&str, ()> = module_ids
        .iter()
        .chain(import_ids.iter())
        .filter(|id| forbidden(id))
        .map(|id| (id.as_str(), ()))
        .collect();
    if offenders.is_empty() {
        return Ok(());
    }
    bail!(
        "route {route}: {kind} reaches host-only module(s) {:?} (W173 server/client boundary lint)",
        offenders.keys().collect::<Vec<_>>()
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    fn strings(items: &[&str]) -> Vec<String> {
        items.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn allows_clean_module_and_import_graph() {
        let module_ids = strings(&["/proj/src/app.tsx", "/proj/src/host/browser.ts"]);
        let import_ids = strings(&["route.chunk-abcdef.js"]);
        assert_no_forbidden_modules("/", "client_entrypoint", &module_ids, &import_ids, browser_forbidden)
            .expect("clean graph must pass");
    }

    #[test]
    fn catches_forbidden_resolved_module_id() {
        // Pre-existing behavior: a forbidden id that rolldown actually
        // resolved and bundled shows up in `module_ids`.
        let module_ids = strings(&["/proj/src/app.tsx", "pg"]);
        let import_ids = strings(&[]);
        let err = assert_no_forbidden_modules("/", "client_entrypoint", &module_ids, &import_ids, edge_forbidden)
            .expect_err("bare `pg` import must be caught");
        assert!(err.to_string().contains("pg"));
    }

    #[test]
    fn catches_forbidden_id_reachable_only_via_externalized_import() {
        // R513-B11 regression: rolldown externalizes `node:*` instead of
        // bundling it, so the offending id never lands in `module_ids` — it
        // only shows up in the chunk's `imports` (our `import_ids`). Before
        // this fix, a hydrate bundle that side-effect-imported `node:fs`
        // (transitively, via the `@mesofact/runtime` barrel) sailed through
        // this lint because `module_ids` alone was clean.
        let module_ids = strings(&["/proj/src/host/browser.ts"]);
        let import_ids = strings(&["route.chunk-abcdef.js", "node:fs", "node:async_hooks"]);
        let err =
            assert_no_forbidden_modules("/", "client_entrypoint", &module_ids, &import_ids, browser_forbidden)
                .expect_err("externalized node:* imports must be caught even when module_ids is clean");
        let msg = err.to_string();
        assert!(msg.contains("node:fs"), "expected node:fs in error, got: {msg}");
        assert!(msg.contains("node:async_hooks"), "expected node:async_hooks in error, got: {msg}");
    }

    #[test]
    fn edge_forbidden_also_scans_import_ids() {
        let module_ids = strings(&[]);
        let import_ids = strings(&["better-sqlite3"]);
        let err = assert_no_forbidden_modules(
            "/edge",
            "ssr placement:\"edge\" entrypoint",
            &module_ids,
            &import_ids,
            edge_forbidden,
        )
        .expect_err("db driver reachable only via import_ids must be caught");
        assert!(err.to_string().contains("better-sqlite3"));
    }
}
